### PR TITLE
added tests for genotype 2 phenotype

### PR DESCRIPTION
**Pull request by bwalsh**
This corresponds with PR for server at https://github.com/ga4gh/server/pull/607 and PR for schemas at https://github.com/ga4gh/schemas/pull/432.   

Thanks for your time.

### CODE REVIEW

**Comment by macieksmuga** [comment]
Thank you @bwalsh! We'll schedule integrating this as soon as we've integrated the reads/references/variants tests for baseline.

**Comment by macieksmuga** [comment]
To amend my previous comment: We need to finish creating the baseline release first. What I meant to say before is that we're excited to discuss how to integrate G2P as soon as we've finished with baseline.

File: ctk-transport/src/main/java/org/ga4gh/ctk/transport/URLMAPPINGImpl.java [hunk 1]
```diff
@@ -72,6 +72,8 @@
         defaultEndpoints.put("ctk.tgt.getVariantSet", "variantsets/{id}");
         defaultEndpoints.put("ctk.tgt.getCallset", "callsets/{id}");
 
+        defaultEndpoints.put("ctk.tgt.searchGenotypePhenotype", "genotypephenotype/search");
+
```
Comment (hjellinek) [inline ctk-transport/src/main/java/org/ga4gh/ctk/transport/URLMAPPINGImpl.java]: Must add the new endpoint to the two `defaulttransport.properties` files, too.

File: ctk-transport/src/main/java/org/ga4gh/ctk/transport/URLMAPPINGImpl.java [hunk 2]
```diff
@@ -413,6 +415,17 @@ public void setGetCallset(String getCallset) {
...
 
     @Override
+    public String getSearchGenotypePhenotype() {
+        return endpoints.get("ctk.tgt.searchGenotypePhenotype");
+    }
+
+    @Override
+    public void setSearchGenotypePhenotype(String searchGenotypePhenotype) {
+        endpoints.put("ctk.tgt.searchGenotypePhenotype", searchGenotypePhenotype);
+
```
Comment (hjellinek) [inline ctk-transport/src/main/java/org/ga4gh/ctk/transport/URLMAPPINGImpl.java]: Unnecessary blank line.

File: ctk-transport/src/main/java/org/ga4gh/ctk/transport/protocols/Client.java [hunk 3]
```diff
@@ -52,6 +52,15 @@
      */
     public final References references = new References();
 
+
+    /**
+     * Provides access to variants-related methods.  For example,
+     * <pre>
```
Comment (hjellinek) [inline ctk-transport/src/main/java/org/ga4gh/ctk/transport/protocols/Client.java]: s/variants/genotype-phenotype/

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.StrictAssertions.fail;
+import static org.ga4gh.cts.api.Utils.catchGAWrapperException;
+
+/**
+ * Tests dealing with searching for datasets.
+ *
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: s/datasets/genotype-phenotypes/

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+ * @author Brian Walsh
+ */
+@Category(DatasetsTests.class)
+public class GenotypePhenotypeSearchIT {
+
+    private static Client client = new Client(URLMAPPING.getInstance());
+
+    /**
+     *  error if no parameters
+     * @throws AvroRemoteException if there's an unanticipated error
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Please use full English sentences, standard capitalization and punctuation in JavaDoc.  I'm not being pedantic - the comments should be as beautiful as the code.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+     */
+    @Test
+    public void ensureCatchNoParametersSearch() throws AvroRemoteException {
+        try {
+            SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+            SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+            fail("should have thrown error");
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Use `Utils.catchGAWrapperException` to catch the exception.  It will automatically fail if no exception is thrown by the thunk it wraps.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        } catch ( Exception e) {
+            // no-op, expected
+        }
+    }
+
+    /**
+     * search for feature
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Better comment, please.  I won't mark the rest of these.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+     * @throws AvroRemoteException
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: JavaDoc produces an obnoxious warning if you don't provide an explanation of _when_ this will throw `AvroException`.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+     */
+    @Test
+    public void simpleFeatureSearch() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setFeature("KIT *wild");
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: `response` can never be null, can it?  The call to `searchGenotypePhenotype` should either throw `GAException` or return a valid `SearchGenotypePhenotypeResponse`.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().get(0)).isNotNull() ;
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: How many objects should `getAssociations` return here?  You should add an assertion for that size before the call to `get(0)`.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    public void simpleEvidenceSearch() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setEvidence("imatinib");
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().get(0)).isNotNull() ;
+    }
+
+
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Please remove extra whitespace.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+    /**
+     * search for Phenotype
+     * @throws AvroRemoteException
+     */
+    @Test
+    public void simplePhenotypeSearch() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setPhenotype("GIST");
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: As above, this should not be `null`.  If it can be null, the `Client.searchGenotypePhenotype` method, or the server endpoint, are behaving in a manner unlike the other `Client`` methods or server endpoints.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    /**
+     * search for Phenotype
+     * @throws AvroRemoteException
+     */
+    @Test
+    public void simplePhenotypeSearchNoFind() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setPhenotype("FooBar");
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Another null test of the response.  I won't mark any others I see below this point, but they should all be fixed.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().size()).isEqualTo(0) ;
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Better:
`assertThat(response.getAssociations()).isEmpty();`

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+
+    /**
+     * search for Feature using external id
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Can you expand on this comment?

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+     * @throws AvroRemoteException
+     */
+    @Test
+    public void simpleFeatureSearchExternalIdentifier() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+
+        ExternalIdentifier id = new ExternalIdentifier() ;
+        id.setDatabase("http://www.monarchinitiative.org/_");
+        id.setIdentifier("CGD:d8c2d551UniProtKB:P10721#P10721-1Region");
+        id.setVersion("*");
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: For these and all subsequent constants that refer to things in your compliance data, please add them to class `TestData`, give them clear, meaningful names, and refer to them in tests by those names.

If you have a constant that's meant _not_ to match something in the compliance data, as when you're trying to provoke an exception, please use a method like `Utils.randomName` or `Utils.randomId`, depending on what's meant, instead of a constant that might accidentally look valid to the casual observer.  It'll make your intention more clear.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        feature.setIds(ids);
+        request.setFeature(feature);
+
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().get(0)).isNotNull() ;
+
+    }
+
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Unnecessary whitespace.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        ExternalIdentifierQuery phenotype = new ExternalIdentifierQuery();
+        phenotype.setIds(ids);
+        request.setPhenotype(phenotype);
+
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().get(0)).isNotNull() ;
+
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Whitespace.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Whitespace.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+        request.setEvidence(evidence);
+
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().get(0)).isNotNull() ;
+
+    }
+
+
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Whitespace.  (I'll stop now. :-))

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+    /**
+     * simple search for Feature, test page size = null
+     * @throws AvroRemoteException
+     */
+    @Test
+    public void testGenotypePhenotypeSearchFeaturePagingMore() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setFeature("KIT *wild");
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Null not possible?

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+    public void testGenotypePhenotypeSearchFeaturePagingAll() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setFeature("KIT *wild");
+        request.setPageSize(1);
+        SearchGenotypePhenotypeResponse response = client.genotypePhenotype.searchGenotypePhenotype(request);
+        assertThat(response).isNotNull();
+        assertThat(response.getAssociations()).isNotNull() ;
+        assertThat(response.getAssociations().size()).isEqualTo(1);
+
+        for(int i = 0 ; i < 3 ; i++) {
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Why is this looping 3 times?  What's special about 3?

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.StrictAssertions.fail;
+import static org.ga4gh.cts.api.Utils.catchGAWrapperException;
+
+/**
+ * Tests dealing with searching for datasets.
+ *
+ * @author Brian Walsh
+ */
+@Category(DatasetsTests.class)
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Wrong category.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+public class GenotypePhenotypeSearchIT {
+
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: __Important__: Where is the additional or modified compliance data that this is supposed to refer to?  I don't see it in this PR - maybe I overlooked it?

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/package-info.java [hunk 25]
```diff
@@ -0,0 +1,4 @@
+/**
+ * This package contains g2p tests, test suite, and test category.
+ */
+package org.ga4gh.cts.api.g2p;
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/package-info.java]: Line ends with EOF.  Please add a newline.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeTestSuite.java [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+import com.googlecode.junittoolbox.IncludeCategories;
+import com.googlecode.junittoolbox.SuiteClasses;
+import com.googlecode.junittoolbox.WildcardPatternSuite;
+
+import org.ga4gh.ctk.testcategories.CoreTests;
+import org.ga4gh.cts.api.datasets.DatasetsTests;
+import org.junit.runner.RunWith;
+
+/**
+ * <p>This suite runs the tests in the "datasets" category.</p>
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeTestSuite.java]: Is `datasets` correct?

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+
+    /**
+     * search for evidence
+     * @throws AvroRemoteException
+     */
+    @Test
+    public void simpleEvidenceSearch() throws AvroRemoteException {
+        SearchGenotypePhenotypeRequest request = SearchGenotypePhenotypeRequest.newBuilder().build();
+        request.setEvidence("imatinib");
```
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: Magic constant?

**Comment by bwalsh** [comment]
@hjellinek   Thanks so much for the thoughtful review.  I've pushed changes onto my branch that address everything.

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java (hunk 24 above)
Comment (bwalsh) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/GenotypePhenotypeSearchIT.java]: I've been unable to get the 'compliance data' to work with the reference server and have commented on existing issues in the server.  The data the server works with is part of the companion PR https://github.com/ga4gh/server/pull/607

File: cts-java/src/test/java/org/ga4gh/cts/api/g2p/package-info.java (hunk 25 above)
Comment (bwalsh) [inline cts-java/src/test/java/org/ga4gh/cts/api/g2p/package-info.java]: Added a new line, for some reason github still shows this line comment

File: ctk-transport/src/main/java/org/ga4gh/ctk/transport/URLMAPPINGImpl.java (hunk 1 above)
Comment (bwalsh) [inline ctk-transport/src/main/java/org/ga4gh/ctk/transport/URLMAPPINGImpl.java]: done

**Comment by sarahhunt** [comment]
Hi @bwalsh ! It tried to test our implementation against the compliance suite, but it looks like the test data we all need to load is missing from the test-data directory.

**Comment by bwalsh** [comment]
@sarahhunt thanks for the feedback!  look for an update early next week.

**Comment by bwalsh** [comment]
closed in favor of https://github.com/ga4gh/compliance/pull/170
